File: matplotlua/matplotlua.lua
-- This module provides functions to add plot objects (points, curves, circles)
-- and generate plots using a Python script. It allows data to be saved to a
-- JSON file or passed directly through a pipe to the Python script.

local cjson = require("cjson")
local Plot = {}

-- we want to call the Python script locally from the same directory
local lib_dir = debug.getinfo(1, 'S').source:match[[^@?(.*[\/])[^\/]-$]]

-- global variables for script and JSON filenames
Plot.SCRIPT_PATH = lib_dir .. "matplotlua.py"
Plot.JSON_NAME = "matplotlua.json"

function Plot.clear()
  Plot.state = {
    points = {},
    curves = {},
    circles = {},
    polygons = {},
  }
end

-- reset plot data and merge with new data
function Plot.init(data)
  Plot.clear()
  for k,v in pairs(data) do Plot.state[k] = v end
end

-- points is a list of tables with x and y coordinates
function Plot.addPointList(points)
  for _, p in ipairs(points) do
    table.insert(Plot.state.points, {
      x = p.x,
      y = p.y,
    })
  end
end

-- x and y can be numbers or lists of numbers
function Plot.addPoint(x, y)
  table.insert(Plot.state.points, {
    x = x,
    y = y,
  })
end

-- x and y are lists of vertices in the curve
function Plot.addCurve(x, y, label)
  table.insert(Plot.state.curves, {
    x = x,
    y = y,
    label = label
  })
end

-- circle is a table with x, y (for the center) and r(adius)
function Plot.addCircle(circle, label, color)
  table.insert(Plot.state.circles, {
    center = {x = circle.x, y = circle.y},
    radius = circle.y,
    label = label,
    color = color,
  })
end

function Plot.addPolygon(points, label, color)
  local vertices = {}
  for _, point in ipairs(points) do
    table.insert(vertices, {point.x, point.y})
  end
  table.insert(Plot.state.polygons, {
    vertices = vertices,
    label = label,
    color = color,
  })
end

-- dump plot data to a file with default or specified name
function Plot.saveToFile(filename)
  filename = filename or Plot.JSON_NAME
  local json_data = cjson.encode(Plot.state)
  local file = io.open(filename, "w")
  file:write(json_data)
  file:close()
end

function Plot.plot(use_file)
  if use_file then
    -- write to a file and call Python script with file
    Plot.saveToFile()
    local command = 'python "' .. Plot.SCRIPT_PATH .. '" ' .. Plot.JSON_NAME
    os.execute(command)
  else
    -- call Python script through a pipe with JSON data
    local json_data = cjson.encode(Plot.state)
    local pipe = io.popen('python "' .. Plot.SCRIPT_PATH .. '"', 'w')
    pipe:write(json_data)
    pipe:close()
  end
end

-- init Plot.state
Plot.clear()

return Plot

File: matplotlua/matplotlua.py
#!/usr/bin/python

# Reads JSON data and plots curves, circles, and points using Matplotlib.

import sys
import json
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon

def plot_from_json(json_data):
    plt.clf()

    for obj in json_data['points']:
        label = obj.get('label') # default to None
        color = obj.get('color') # default to None
        plt.plot(obj['x'], obj['y'], 'o', label=label)
    for obj in json_data['curves']:
        label = obj.get('label')
        color = obj.get('color')
        plt.plot(obj['x'], obj['y'], label=label)
    for obj in json_data['polygons']:
        label = obj.get('label')
        color = obj.get('color')
        vertices = obj['vertices']
        polygon = Polygon(vertices, closed=True, fill=False, color=color, label=label)
        plt.gca().add_patch(polygon)
    for obj in json_data['circles']:
        label = obj.get('label')
        color = obj.get('color')
        center = obj['center']
        radius = obj['radius']
        circle = plt.Circle((center['x'], center['y']), radius, color=color, label=label, fill=False)
        plt.gca().add_patch(circle)

    plt.legend()
    plt.xlabel(json_data['xlabel'] if 'xlabel' in json_data else 'X')
    plt.ylabel(json_data['ylabel'] if 'ylabel' in json_data else 'Y')
    plt.title(json_data['title'] if 'title' in json_data else 'MatPlotLua')

    if 'figure' in json_data:
        plt.savefig(json_data["figure"])
    else:
        plt.show()

if __name__ == '__main__':
    if len(sys.argv) == 2:
        # example: python matplotlua.py example.json
        json_filename = sys.argv[1]
        with open(json_filename, 'r') as json_file:
            data = json.load(json_file)
            plot_from_json(data)
    else:
        # if no argument is provided, read JSON data from stdin
        # that's useful for piping data from another program
        # example: cat example.json | python matplotlua.py
        json_data = sys.stdin.read()
        data = json.loads(json_data)
        plot_from_json(data)
